### DISCUSSION TITLE
Smarter export

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12275] [Ideas]
## Problem

In theory Dagger is perfect for generating code or docs. In practice, the logic for exporting files back to the client filesystem is simplistic and brittle, which makes the experience awkward for those use cases.

### Limitations of `export`

1. User must always specify the target path to write to. For example `dagger call codegen -o .`
2. The target path is relative to the working directory, which means the command must be changed to match the current workdir - this makes the use of Dagger brittle. For example: `cd foo; dagger call codegen -o ..; cd bar; dagger call codegen -o ../..`
3. `export` can only add files by default, not remove them. For example it can't remove previously generated files if they become out of date.
4. Files can be removed with `export(wipe: true)`, but it is very easy to remove *too much*. In combination with the mandatory `path`, it is a recipe for disastrous user errors. Ask @sipsma 😛

### User reports

- [RoffleeCopterz asks about scaffolding](https://discord.com/channels/707636530424053791/1277646308340928512)
- [shykes asks about generating CI config](https://discord.com/channels/707636530424053791/1275901633880129749/1275910866403070045)

## Solution

Implement first-class support for codegen pipelines, in a way that doesn't break the fundamentals of the Dagger API.

## Design

Make `export` smarter by making it *sync changes* instead of just dumping content.

Here's how it would work:

1. All filesystem changes are synced to the target: written files and directories, as well as *removed* files and directories.
2. No other content is written to the target. For example, if a file is present in the exported directory, but hasn't been created or changed during this run, it's ignored.
3. The default target path is the source
4. If the source isn't known, or can't be written to, the default target causes an error

#### Adding a file

```
host |
directory . |
with-new-file src/hello.txt 'hello world' |
export
```

#### Removing a file

```
host |
directory .  |
without-file src/hello.txt |
export
```

#### Renaming a directory

```
old="$(host | file old.txt)"
host |
directory . |
without-file old.txt |
with-file new.txt $old |
export
```

#### Codegen then export back to context

```golang
func (m *Mod) Generate(
  // +defaultPath="."
  source *dagger.Directory,
) *dagger.Directory {
  return source.
    WithoutFile("foo").
    WithNewFile("bar", "bla bla bla")
}
```

```console
generate | export
```

### Design considerations

The main consideration is implementation difficulty. To properly sync changes back to the source, one must 1) know what the changes are, and 2) know what the source is.

> know what the changes are

Most operations are deterministic: `withFile`, `withoutFile`, etc.

But there is also `withExec`... This requires comparing the rootfs before and after execution, to add the corresponding changes to the sync.

> know what the source is

This may or may not be tricky. For example:

```console
directory |
with-directory foo $(host | directory .) |
with-new-file foo/bar "hello" |
directory foo |
export
```

In this example, the source of the exported directory is `./foo`.

**Comment by shykes** [comment]
FYI @aluzzardi @sipsma @vito as discussed previously

**Comment by shykes** [comment]
Relevant to #9079

**Comment by shykes** [comment]
@marcosnils 👋 I consolidated #9079 into this one, which predates it.

**Comment by sagikazarmark** [comment]
I don't know if it changes the scope, but there is another similar use case: code formatting.

The primary difference is that generated code is usually not modified by hand (or other tools, like code formatters) whereas handwritten code may be.

Another potential difference is generated code is often in a single or a few places within a codebase where nothing else lives which could simplify exporting content (eg. overwrite everything in a folder). That's not the case with "other" files that formatters may touch.

Personally, I've been _mostly_ satisfied with how Dagger currently works for code generation use cases (I usually just overwrite an entire generated folder). But I haven't really used Dagger for code formatting (yet).

**Comment by shykes** [comment]
Yes definitely, code formatting is another use case in this category. I'm not sure if there's an appropriate name for the overall category that encompasses codegen, docs gen, config file generation, and code formatting?

**Comment by sipsma** [comment]
The hardest problem here lies around caching; specifically in magically tracking the source of a directory to support a default export target. These directories synced from the host are cached based on their content, completely disconnected from the paths/clients they were sourced from. This is important to preserve to keep our cache reusability maximal.

If we made the synced in Directory include state on where it was sourced from we'd either:
1. Have to include that path in the cache key, which breaks the cache reusability mentioned above
2. Not include that path in the cache key, which creates extremely confusing situations where you end up getting cached return values that don't have the same "client source directory" embedded in them. This would just be inconsistent+broken, not a viable option
3. Something very convoluted where we unset the "client source directory" before passing as an arg to a function, then attempt to find any references to it in the return value and reset it
   * I'm not even sure if this is actually possible in every case, but *maybe* in the cards. Would without a doubt be another source of convolution in the code, though that's admittedly just an internal concern as opposed to user facing

I guess `1.` is an option but feels like we'd be permanently giving up something really fundamental that'd we'd regret eventually (or immediately).

`2.` isn't even worth considering IMO.

I wouldn't discount `3.` entirely, but will mention what an alternative that avoids all this looks like.

These problems would go away if the UX changed so that the user had to provide the "original" source as an arg to export. Using rough shell pseudocode and a hypothetical new `--diff` option on `export`, something like:

```
SRCDIR=$(host | directory .)
$SRCDIR |
with-new-file src/hello.txt 'hello world' |
export --diff $SRCDIR
```

It's obviously not as succinct, especially for very simple cases like that example but would:
1. Avoid all the issues mentioned above
2. Be more explicit and clear, which might be a benefit for less straightforward cases where the directory goes through all sorts of transformations before being exported back to the host (this is debatable)
3. Be a bit more generic since the argument to `--diff` can theoretically be *any* Directory arg, not just local dirs. Admittedly I can't think of super compelling examples for this yet, so might be irrelevant.

Just fishing for thoughts from others here.

**Comment by shykes** [comment]
Note: now that Dagger can run dev environments for coding agents, this problem is even more of a blocker.

**Comment by ajw0100** [comment]
+1 for the codegen, docs gen, config file generation, and code formatting use cases. In addition to using dagger to orchestrate CI workflows, I want to use it to provision the local dev toolchain and orchestrate local dev workflows in a manner that is hermetic, reproducible, etc. Similar to using Nix to build local dev environemnts but less strict, better ergonomics, gentler learning curve.

**Comment by vito** [comment]
Been experimenting with Dagger native coding agents lately. Ran into this and so far I've landed on this scheme:

* There's a "current environment" (`Env`) accessed via `Query.currentEnv`
* `Env` has a `hostfs` (part of #10370)
  * In a module, `currentEnv.hostfs` maps directly to `ModuleSource.contextDirectory`
  * In an LLM session, `currentEnv` is directly the `LLM.env`, and has whatever hostfs it was set up with, possibly from `Query.currentEnv` when the agent was created
* `Env` can have modules installed into it (`Env.withModule(Module)`)
  * For an LLM, installed modules have their tools directly exposed to the agent
    * When calling these module tools, the `Env.hostfs` is used to populate all context path args, and any returned objects (`DaggerDev#1`) are implicitly re-instantiated with the latest `Env.hostfs` on use
    * This is a lot of words to say "Env hostfs changes automatically propagate to future module calls even on chained functions"
  * In the future, when `Env` can be used to evaluate Dagger shell (another part of #10370), these modules determine the shell context (i.e. the schema)
* Functions return an `Env` to signal their intent to update the caller's environment
  * When an LLM calls a function that returns an `Env`, that `Env` becomes `LLM.env`, and it propagates to future module calls
  * When the Dagger CLI calls a function that returns an `Env`, it can prompt or automatically (if safe?) sync the changes to the caller

Example pattern: `binary: File!` which returns a compiled binary, and `build: Env!` that exports the `binary` to `./bin/dagger`. You would use `binary` in code, and `build` from the CLI or coding agent. `generate: Env!` would be another common example.

Parts I like:

* Clear intent when the goal of a function is to export (return `Env`)
* Having an `Env` before and after means we can do a diff and only export the changed files. (So users don't have to implement the diffing themselves anymore, just return a full `Env` and let the diffing happen externally.)
* LLM semantics seem pretty clear (return `Env`) - previously `Env` was a banned type, now it has high level meaning

Parts I'm "meh" on:

* API redundancy between `binary: File!` and `build: Env!` - but, their intended use _is_ different, and they're cheap to maintain (`Env` variant is trivial extension of non-`Env` variant), so maybe it's OK?

Gaps:

* Diffing magic doesn't work for deleted files.
  * Can kind of do this at the application layer though: glob `**/*` in Before and After, and `rm -rf` any paths that only appear in Before.

**Comment by jedevc** [comment]
I think we can potentially remove this given the existence of `Changeset`s?
